strutil/quantity/example_test.go

```diff
@@ -115,3 +115,2866 @@ func ExampleFormatDuration() {
 	// " 18Ey"
 	// "ages!"
 }
+
+var dataset = []float64{
+	-math.MaxFloat64,
+	float64(math.MinInt64),
+	-time.Duration(3 * time.Nanosecond).Seconds(),
+	time.Duration(3 * time.Nanosecond).Seconds(),
+	time.Duration(36 * time.Microsecond).Seconds(),
+	time.Duration(36*time.Microsecond + 3*time.Nanosecond).Seconds(),
+	time.Duration(420*time.Millisecond + 36*time.Microsecond + 3*time.Nanosecond).Seconds(),
+	time.Duration(430 * time.Millisecond).Seconds(),
+	time.Duration(5155 * time.Millisecond).Seconds(),
+	time.Duration(time.Minute + 2*time.Second).Seconds(),
+	time.Duration(124 * time.Minute / 10).Seconds(),
+	time.Duration(2*time.Hour + 29*time.Minute).Seconds(),
+	time.Duration(10*time.Hour + 9*time.Minute).Seconds(),
+	time.Duration(10*time.Hour + 30*time.Minute).Seconds(),
+	time.Duration(10*time.Hour + 30*time.Minute + 2*time.Second).Seconds(),
+	time.Duration(11*time.Hour + 2*time.Minute).Seconds(),
+	time.Duration(22*time.Hour + 59*time.Minute + 59*time.Second).Seconds(),
+	time.Duration(23*time.Hour + 59*time.Minute + 59*time.Second).Seconds(),
+	time.Duration(30 * time.Hour).Seconds(),
+	time.Duration(345 * time.Hour).Seconds(),
+	time.Duration(357 * time.Hour).Seconds(),
+	time.Duration(4272 * time.Hour).Seconds(),
+	time.Duration(4273*time.Hour + 30*time.Minute + 2*time.Second).Seconds(),
+	time.Duration(51368 * time.Hour).Seconds(),
+	time.Duration(math.MaxInt64 / 10).Seconds(),
+	time.Duration(math.MaxInt64).Seconds(),
+	float64(math.MaxUint64) * 365 * 24 * 60 * 60,
+	math.MaxFloat64,
+}
+
+func Inv(dt float64) {
+	fmt.Printf("%q\n", quantity.FormatDurationGeneric(dt,
+		quantity.Year,
+		quantity.Sec,
+		quantity.ShowAll,
+		quantity.TimeLeft,
+		true))
+}
+
+func ExampleInv() {
+
+	for _, dt := range dataset {
+		Inv(dt)
```

Comment:
This is not your fault at all, but I think before we can merge this PR, we really need to clean up this test file because it is not matching go-check at all, and AFAICT most of these functions that are already in this package are not even being called by any test infrastucture we have? :confused:

Reply:
Indeed, instead of printing the output, we should actually compare it with the expected one and fail the test if they don't match.

Reply:
I think there is a misunderstanding here. All code in the test is 100% tested.

The Example Test rules dictate that ExampleXXX should prefix the function name.

However, the base function is a generic renderer, with many permutations to test. Instead of create an "Output:" section with all the data in one long list, I logically grouped the tests.

Will reach out to discuss.

Reply:
My question is resolved (and was due to my own ignorance indeed). I'll leave this one unresolved until we hear Ian's point too.

Reply:
Snapd uses the "go check" (gopkg.in/check.v1) unit test framework which is bolted on top of the built-in Go unit test framework "testing". 

The hook is installed in each package (func Test(t *testing.T) { TestingT(t) }), which allows the "go test ." mechanism to just work, calling the go check Test Suites registered in the package.

The "strutil/quantity" duration string renderer uses the Go built-in unit test "Example*" framework in order to efficiently verify the entire range of string rendering modes and options, logically grouped in the unit test file, by comparing stdout against output defined in the unit test file ("Output: ...").

Both the built-in and go check framework is automatically called when "go test ." is invoked. This is done from the "run-checks" script. 

In the spread testing environment, unit tests are defined under "tests/unit/go/task.yaml". This task includes both "run-checks --static" as well as "run-checks --unit", were the latter will ultimately call "go test xxx".

Reply:
We discussed this in the SU, and while we are not opposed to having some tests in this Example... format, we don't think it's appropriate for _all_ tests here to be in the Example format, and would prefer to have almost all of the functional tests to be written with go-check. So it's probably okay to leave the existing Example... tests, but I think the new tests which you have added should be put into the go-check format, simply because many of the tests you have added are not really indicative of how to use this package but are more functional in ensuring that the behavior is correct for corner cases. For example the following tests all seem to be corner cases and not really good "example" code:

* ExampleSpaceAllTimeLeftRange1
* ExampleSpaceAllTimeLeftRange2
* ExampleSpaceAllTimeLeftRange3

etc

Does this make sense? 

In general, if we want to have Example... functions, it probably makes sense to just have 1 or 2 examples per exported function, but here all of the tests are for one singular exported function, quantity.FormatDurationGeneric.

Reply:
I will do it, no problem.

Reply:
All tests (FormatDurationGeneric) moved to quantity_test.go (go check framework), apart from the two specific function examples (ProgressBarTimePassed, ProgressBarTimeLeft).

100% coverage of new code.

Resolved.